README.md
[![Build Status](https://travis-ci.org/Jaguar-dart/jaguar_serializer_mongo.svg?branch=master)](https://travis-ci.org/Jaguar-dart/jaguar_serializer_mongo)

# jaguar_serializer_mongo

Mongo serializer helpers for [jaguar_serializer](https://github.com/Jaguar-dart/jaguar_serializer)

## Simple example

```dart
import 'package:mongo_dart/mongo_dart.dart';
import '../models/player/player.dart';

main() async {
  Db db = new Db('mongodb://localhost/jaguar_mongo_test');
  await db.open();

  try {
    DbCollection coll = db.collection('one');
    ObjectId id = new ObjectId();
    ObjectId allianceId = new ObjectId();

    Player player = new Player()
      ..id = id.toHexString()
      ..allianceId = allianceId.toHexString()
      ..name = 'KnightsOT'
      ..email = 'knightsot@example.com'
      ..age = 27
      ..score = 500000
      ..emailConfirmed = true;

    await coll.insert(Player.serializer.toMap(player));

    Map result = await coll.findOne(where.id(id));
    print(result);

    Player decoded = Player.serializer.fromMap(result);
    print(decoded);
  } finally {
    await db.drop();
    await db.close();
  }
}
```

CHANGELOG.md
# Changelog

## 0.2.0

- Moved to jaguar_serializer 0.5.x

## 0.1.0

- Moved to jaguar_serializer 0.4.x

## 0.0.4

- Added MgoAccess

## 0.0.3

- Upgraded to jaguar_serializer 0.3.x

## 0.0.1

- Initial version, created by Stagehand

test/test_all.dart
import 'simple/simple_test.dart' as simple;

void main() {
  simple.main();
}

test/simple/simple_test.dart
// Copyright (c) 2016, Ravi Teja Gudapati. All rights reserved. Use of this source code
// is governed by a BSD-style license that can be found in the LICENSE file.

import 'package:mongo_dart/mongo_dart.dart';
import 'package:test/test.dart';
import '../models/player/player.dart';

main() {
  group('mongo.ObjectId', () {
    Db db;

    setUp(() async {
      db = new Db('mongodb://localhost:27018/jaguar_mongo_test');
      await db.open();
    });

    tearDown(() async {
      await db.drop();
      await db.close();
    });

    test('Encode', () async {
      DbCollection coll = db.collection('one');
      ObjectId id = new ObjectId();
      ObjectId allianceId = new ObjectId();

      Player player = new Player()
        ..id = id.toHexString()
        ..allianceId = allianceId.toHexString()
        ..name = 'KnightsOT'
        ..email = 'knightsot@example.com'
        ..age = 27
        ..score = 500000
        ..emailConfirmed = true;

      await coll.insert(Player.serializer.toMap(player));

      Map result = await coll.findOne(where.id(id));
      expect(result['_id'], new isInstanceOf<ObjectId>());
      expect(result['_id'], id);

      expect(result['allianceId'], new isInstanceOf<ObjectId>());
      expect(result['allianceId'], allianceId);

      expect(result['name'], 'KnightsOT');
      expect(result['email'], 'knightsot@example.com');
      expect(result['age'], 27);
      expect(result['score'], 500000);
      expect(result['emailConfirmed'], isTrue);
    });

    test('Decode', () async {
      DbCollection coll = db.collection('one');
      ObjectId id = new ObjectId();
      ObjectId allianceId = new ObjectId();

      Player player = new Player()
        ..id = id.toHexString()
        ..allianceId = allianceId.toHexString()
        ..name = 'KnightsOT'
        ..email = 'knightsot@example.com'
        ..age = 27
        ..score = 500000
        ..emailConfirmed = true;

      await coll.insert(Player.serializer.toMap(player));

      Map result = await coll.findOne(where.id(id));
      Player decoded = Player.serializer.fromMap(result);

      expect(decoded.id, new isInstanceOf<String>());
      expect(decoded.id, id.toHexString());

      expect(decoded.allianceId, new isInstanceOf<String>());
      expect(decoded.allianceId, allianceId.toHexString());

      expect(decoded.name, 'KnightsOT');
      expect(decoded.email, 'knightsot@example.com');
      expect(decoded.age, 27);
      expect(decoded.score, 500000);
      expect(decoded.emailConfirmed, isTrue);
    });
  });
}


test/models/player/player.dart
library serializer.test.models.player;

import 'package:jaguar_serializer/jaguar_serializer.dart';
import 'package:jaguar_serializer_mongo/jaguar_serializer_mongo.dart';

part 'player.g.dart';

@GenSerializer(fields: const {
  'id': const EnDecode(alias: '_id', processor: const MongoId()),
  'allianceId': const EnDecode(processor: const MongoId()),
})
class PlayerSerializer extends Serializer<Player> with _$PlayerSerializer {
  Player createModel() => new Player();

  PlayerSerializer();
}

/// Player model for the game
class Player {
  /// Id of the player
  String id;

  /// Id of the alliance the player belongs to
  String allianceId;

  /// Name of the player
  String name;

  /// Email of the player
  String email;

  /// Age of the player
  int age;

  /// Player score
  int score;

  /// Has the player confirmed his email?
  ///
  /// Should not be sent to client
  bool emailConfirmed;

  static PlayerSerializer serializer = new PlayerSerializer();

  String toString() =>
      [id, allianceId, name, email, age, score, emailConfirmed].toString();
}


lib/src/serializers.dart
// Copyright (c) 2016, Ravi Teja Gudapati. All rights reserved. Use of this source code
// is governed by a BSD-style license that can be found in the LICENSE file.

library serializer.mongo_serializer;

import 'package:mongo_dart/mongo_dart.dart' as mgo;
import 'package:jaguar_serializer/jaguar_serializer.dart';

/// Field processor to serializer Mongo ObjectId between Dart Object and Mongo
class MongoId implements FieldProcessor<String, mgo.ObjectId> {
  const MongoId();

  /// Called to process field before decoding
  String deserialize(mgo.ObjectId input) {
    return input.toHexString();
  }

  /// Called to process field before encoding
  mgo.ObjectId serialize(String value) {
    return new mgo.ObjectId.fromHexString(value);
  }
}

/// Field processor to serializer Mongo DateTime between Dart Object and Mongo
class MongoDateTime implements FieldProcessor<DateTime, DateTime> {
  const MongoDateTime();

  /// Called to process field before decoding
  DateTime deserialize(DateTime value) => value;

  /// Called to process field before encoding
  DateTime serialize(DateTime value) => value;
}

abstract class MgoAccess<ModelType> {
  Serializer<ModelType> get codec;

  ModelType decode(Map data) {
    return codec.fromMap(data);
  }

  List<ModelType> decodeList(List<Map> data) {
    return data.map((Map item) => codec.fromMap(item)).toList();
  }

  Map encode(ModelType data) {
    return codec.toMap(data);
  }

  List<Map> encodeList(List<ModelType> data) {
    return data.map((ModelType item) => codec.toMap(item)).toList();
  }
}


example/models/player/player.dart
library serializer.test.models.player;

import 'package:jaguar_serializer/serializer.dart';
import 'package:jaguar_serializer_mongo/serializers.dart';

part 'player.g.dart';

@EnDecodeField(#id, asAndFrom: '_id')
@MongoId(#id)
@MongoId(#allianceId)
@GenSerializer()
class PlayerSerializer extends _$PlayerSerializer
    implements MapSerializer<Player> {
  Player createModel() => new Player();

  PlayerSerializer();
}

/// Player model for the game
class Player {
  /// Id of the player
  String id;

  /// Id of the alliance the player belongs to
  String allianceId;

  /// Name of the player
  String name;

  /// Email of the player
  String email;

  /// Age of the player
  int age;

  /// Player score
  int score;

  /// Has the player confirmed his email?
  ///
  /// Should not be sent to client
  bool emailConfirmed;

  static PlayerSerializer serializer = new PlayerSerializer();

  String toString() =>
      [id, allianceId, name, email, age, score, emailConfirmed].toString();
}


example/models/player/player.g.dart
// GENERATED CODE - DO NOT MODIFY BY HAND

part of serializer.test.models.player;

// **************************************************************************
// Generator: SerializerGenerator
// Target: class PlayerSerializer
// **************************************************************************

abstract class _$PlayerSerializer implements MapSerializer<Player> {
  const _$PlayerSerializer();

  Map toMap(Player model) {
    Map ret = new Map();
    ret["_id"] = new MongoId(#id).to(model.id);
    ret["allianceId"] = new MongoId(#allianceId).to(model.allianceId);
    ret["name"] = model.name;
    ret["email"] = model.email;
    ret["age"] = model.age;
    ret["score"] = model.score;
    ret["emailConfirmed"] = model.emailConfirmed;
    return ret;
  }

  Player fromMap(Map map, {Player model}) {
    if (map is! Map) {
      return null;
    }
    if (model is! Player) {
      model = createModel();
    }
    model.id = new MongoId(#id).from(map["_id"]);
    model.allianceId = new MongoId(#allianceId).from(map["allianceId"]);
    model.name = map["name"];
    model.email = map["email"];
    model.age = map["age"];
    model.score = map["score"];
    model.emailConfirmed = map["emailConfirmed"];
    return model;
  }
}


example/simple/simple_example.dart
library jaguar_serializer_mongo.example_simple;

import 'package:mongo_dart/mongo_dart.dart';
import '../../test/models/player/player.dart';

main() async {
  Db db = new Db('mongodb://localhost:27018/jaguar_mongo_test');
  await db.open();

  try {
    DbCollection coll = db.collection('one');
    ObjectId id = new ObjectId();
    ObjectId allianceId = new ObjectId();

    Player player = new Player()
      ..id = id.toHexString()
      ..allianceId = allianceId.toHexString()
      ..name = 'KnightsOT'
      ..email = 'knightsot@example.com'
      ..age = 27
      ..score = 500000
      ..emailConfirmed = true;

    await coll.insert(Player.serializer.toMap(player));

    Map result = await coll.findOne(where.id(id));
    print(result);

    Player decoded = Player.serializer.fromMap(result);
    print(decoded);
  } finally {
    await db.drop();
    await db.close();
  }
}
